bond_app/bond.py

```diff
@@ -60,23 +62,76 @@ def exchange_authz_code(self, authz_code, redirect_uri, sam_user_id):
         if refresh_token is not None:
             token_response = self.oauth_adapter.refresh_access_token(refresh_token.token)
             expires_at = datetime.fromtimestamp(token_response.get(FenceKeys.EXPIRES_AT))
+
             return token_response.get("access_token"), expires_at
         else:
             raise exceptions.NotFound(
                 "Could not find refresh token for sam_user_id: {} provider_name: {}\nConsider relinking your account to Bond.".format(
                     sam_user_id, self.provider_name))
 
+
+    def get_access_token(self, sam_user_id, refresh_threshold: int = 600):
+        """
+        Given a user, lookup their refresh token (if not provided) and use it to retrieve an access token from their OAuth
+        provider.
+        If an access token was already generated for the user, 
+        and that token has greater than `refresh_threshold` seconds before expiration, return that token. 
+        Otherwise, generate a new token.
+        
+        If a refresh token cannot be found for the sam_user_id provided, a NotFound will be raised.
+        :param sam_user_id: Id stored in Sam for user who initiated request
+        :return: Two values: An Access Token string, datetime when that token expires
+        """
+        refresh_token = self.refresh_token_store.lookup(sam_user_id, self.provider_name)
+        if refresh_token is not None:
+            cached_access_data = self.cache_api.get(namespace="AccessTokens", key=sam_user_id)
+            if cached_access_data and cached_access_data.get(FenceKeys.ACCESS_TOKEN):
+                expires_at = cached_access_data.get(FenceKeys.EXPIRES_AT)
+                access_token = cached_access_data.get(FenceKeys.ACCESS_TOKEN)
+                expires_in_cache = (expires_at - datetime.now()).total_seconds() - refresh_threshold
+                logging.debug(
+                    "Retrieved access token from cache. " +
+                    f"Access token will expire in {expires_in_cache:.2f} seconds. " +
+                    f"SAM user ID: {sam_user_id}. Provider: {self.provider_name}"
+                )
+            else:
+                access_token, expires_at = self.generate_access_token(sam_user_id, refresh_token=refresh_token)
+                expires_in_cache = (expires_at - datetime.now()).total_seconds() - refresh_threshold
+                logging.debug(
+                    "Generated new access token. " +
+                    f"Access token will expire in {expires_in_cache:.2f} seconds. " +
+                    f"SAM user ID: {sam_user_id}. Provider: {self.provider_name}"
+                )
+                
+                self.cache_api.add(
+                    namespace="AccessTokens", 
+                    key=sam_user_id, 
+                    value={
+                        FenceKeys.EXPIRES_AT: expires_at,
+                        FenceKeys.ACCESS_TOKEN: access_token,
```

Comment:
`FenceKeys` is for a different use case, I would not reuse them here. I would create a class to hold a Fence access token instead. It reads better that dealing with a dict.

Reply:
would you put the Fence access token class here in `bond.py`? or elsewhere?